### PR TITLE
Include types for 'locale' module from fusion-plugin-i18n

### PR DESCRIPTION
**Pull request by AlexMSmithCA**
I'd like thoughts on three approaches for exporting types for third-party packages.  In an ideal world, the packages themselves are either typed or expose libdef files through [flow-typed](https://github.com/flow-typed/flow-typed).  In this case, however, we have created our own type definitions for `locale` as stored them in [`fusion-plugin-i18n's` `flow-typed` directory](https://github.com/fusionjs/fusion-plugin-i18n/blob/master/flow-typed/locale.js).

**Approach 1**: Ship third-party module types in `flow-typed/` directory as libdef files
- Pro: we have precedent for doing this.  We do this for `tape-cup` in [`fusion-core`](https://github.com/fusionjs/fusion-core/tree/master/flow-typed).
- Con: packages have to ship third-party module types.
- Con: not commonly used outside of our codebase.
- Con: requires consumers to [include `flow-typed/` folder of dependencies in `.flowconfig`](https://github.com/fusionjs/fusion-plugin-i18n/blob/master/.flowconfig#L9) to utilize types.

**Approach 2**: Ship third-party module types in `src/` directory as normally exported types
- Pro: consumers do not have to change their `.flowconfig` to include these types.
- Pro: package exports that are typed with these types will have their types flow through to consumers.
- Con: packages have to ship third-party module types.
- Con: types will not be available if consumers import from the third-party packages directly.

**Approach 3**: Publish third-party types to [flow-typed](https://github.com/flow-typed/flow-typed) and ensure consumers install types.
- Pro: adopted as a best standard by community.
- Pro: contributing to an open source project to improve overall type coverage within Flow community.
- Con: additional (manual?) package installation step to download third-party types (e.g. "if you use `fusion-plugin-i18n`, you must also download types for `locale`). 
- Con: our types are often subsets and specific to our use cases.
- Con: slower iteration cycle for changing these types if we rely on a third-party (compared to publishing new types in our normal release pipeline).

### CODE REVIEW

File: .flowconfig [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 [libs]
 ./node_modules/fusion-core/flow-typed
+./node_modules/fusion-plugin-i18n/flow-typed
```
Comment (AlexMSmithCA) [inline .flowconfig]: Consumers would have to add this to their `.flowconfig` to include the `libdef` for [`locale`](https://github.com/fusionjs/fusion-plugin-i18n/blob/master/flow-typed/locale.js).  Without this, imports would be of type `any` (and are consequently uncovered).

File: package.json [hunk 2]
```diff
@@ -31,7 +31,7 @@
   },
   "dependencies": {
     "fusion-cli": "^1.12.1",
-    "fusion-plugin-i18n": "^1.1.4",
+    "fusion-plugin-i18n": "1.1.6-0",
```
Comment (AlexMSmithCA) [inline package.json]: (Pre-release includes `flow-typed/` folder in npm package)

**Comment by ganemone** [comment]
Is there a precedent for how this is done in the open source community?

**Comment by AlexMSmithCA** [comment]
I don't think **Approach 2** is viable in this case.  As far as I know, Flow only picks up module declarations (e.g. `declare module 'locale' { ... }`) in the `flow-typed/` directory.

Given that, properly typing a third party module would require something like:
```
<src/index.js>
import {Locale} from 'locale';
import type {LocaleType} from './locale-types.js';
const typedLocale = ((Locale: any): LocaleType);
export {typedLocale as Locale};
```

**Comment by AlexMSmithCA** [comment]
I don't think **Approach 2** is a great plugin-author experience in this case.  As far as I know, Flow only picks up module declarations (e.g. `declare module 'locale' { ... }`) in the `flow-typed/` directory.  Placing types in `src/` instead would require we manually coerce types for anything coming from the third party module.

Using `locale` as an example, we would have to have code that re-types any exports that come from the third-party module.

```
<src/locale-types.js>
declare class LocaleCls { ... }; // would otherwise live in libdef in flow-typed directory

<src/some-other-file.js>
import {Locale as LocaleUntyped} from 'locale';
const Locale = ((LocaleUntyped: any): LocaleCls;
// we could then export Locale and consumers would pick up types defined in `src/locale-types.js`
```

This would have to be done anywhere we have an import from third-party packages.

**Comment by AlexMSmithCA** [comment]
@ganemone - **Approach 3** is the de facto "best practice" approach for third-party library definitions.  In this case, it would likely be sufficient since `locale` is open-source and we could contribute to the [flow-typed](https://github.com/flow-typed/flow-typed) project.

This does not solve the same issue for non-OSS projects that do not ship types themselves.

**Comment by AlexMSmithCA** [comment]
!merge

**Pull request merged**